File: sdk/typescript/src/wallet.ts
/**
 * Wallet Manager for Gillean Blockchain TypeScript SDK
 */

import { APIClient } from './client';
import { WalletInfo, WalletCreationOptions, GilleanError } from './types';

/**
 * Manages wallet operations for the Gillean blockchain
 */
export class WalletManager {
  private apiClient: APIClient;

  constructor(apiClient: APIClient) {
    this.apiClient = apiClient;
  }

  /**
   * Create a new wallet
   */
  async createWallet(options: WalletCreationOptions = {}): Promise<WalletInfo> {
    const data = {
      name: options.name || `Wallet_${Date.now()}`,
      password: options.password || 'default_password',
      encryption_enabled: options.encryptionEnabled !== false
    };

    return this.apiClient.post<WalletInfo>('/wallet/create', data);
  }

  /**
   * Get wallet information
   */
  async getWallet(address: string): Promise<WalletInfo> {
    return this.apiClient.get<WalletInfo>(`/wallet/${address}`);
  }

  /**
   * Get all wallets
   */
  async getAllWallets(): Promise<WalletInfo[]> {
    return this.apiClient.get<WalletInfo[]>('/wallet/all');
  }

  /**
   * Get wallet balance
   */
  async getBalance(address: string): Promise<number> {
    const response = await this.apiClient.get<{ balance: number }>(`/wallet/${address}/balance`);
    return response.balance;
  }

  /**
   * Import wallet from private key
   */
  async importWallet(privateKey: string, name?: string): Promise<WalletInfo> {
    const data = {
      private_key: privateKey,
      name: name || `Imported_Wallet_${Date.now()}`
    };

    return this.apiClient.post<WalletInfo>('/wallet/import', data);
  }

  /**
   * Export wallet private key
   */
  async exportPrivateKey(address: string, password: string): Promise<string> {
    const data = { password };
    const response = await this.apiClient.post<{ private_key: string }>(`/wallet/${address}/export`, data);
    return response.private_key;
  }

  /**
   * Delete wallet
   */
  async deleteWallet(address: string, password: string): Promise<void> {
    const data = { password };
    await this.apiClient.delete(`/wallet/${address}`, { data });
  }

  /**
   * Update wallet name
   */
  async updateWalletName(address: string, newName: string): Promise<WalletInfo> {
    const data = { name: newName };
    return this.apiClient.put<WalletInfo>(`/wallet/${address}/name`, data);
  }

  /**
   * Change wallet password
   */
  async changePassword(address: string, oldPassword: string, newPassword: string): Promise<void> {
    const data = {
      old_password: oldPassword,
      new_password: newPassword
    };
    await this.apiClient.put(`/wallet/${address}/password`, data);
  }

  /**
   * Backup wallet
   */
  async backupWallet(address: string, password: string): Promise<string> {
    const data = { password };
    const response = await this.apiClient.post<{ backup_data: string }>(`/wallet/${address}/backup`, data);
    return response.backup_data;
  }

  /**
   * Restore wallet from backup
   */
  async restoreWallet(backupData: string, password: string): Promise<WalletInfo> {
    const data = {
      backup_data: backupData,
      password
    };
    return this.apiClient.post<WalletInfo>('/wallet/restore', data);
  }

  /**
   * Get wallet transaction history
   */
  async getTransactionHistory(address: string, limit: number = 50, offset: number = 0): Promise<any[]> {
    return this.apiClient.get<any[]>(`/wallet/${address}/transactions`, {
      params: { limit, offset }
    });
  }

  /**
   * Validate wallet address
   */
  async validateAddress(address: string): Promise<boolean> {
    try {
      await this.apiClient.get(`/wallet/validate/${address}`);
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get wallet statistics
   */
  async getWalletStats(address: string): Promise<any> {
    return this.apiClient.get(`/wallet/${address}/stats`);
  }
}

File: sdk/src/wallet.rs
use super::{SDKResult, SDKError, SDKConfig, WalletInfo};
use ed25519_dalek::{SigningKey, VerifyingKey};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use sha2::Digest;
use rand::RngCore;

/// Wallet manager for creating and managing wallets
pub struct WalletManager {
    _config: SDKConfig,
    _wallets: HashMap<String, WalletData>,
}

/// Internal wallet data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
struct WalletData {
    address: String,
    name: Option<String>,
    encrypted_private_key: Vec<u8>,
    public_key: Vec<u8>,
    created_at: i64,
    balance: f64,
}

impl WalletManager {
    /// Create a new wallet manager
    pub fn new(config: SDKConfig) -> Self {
        Self {
            _config: config,
            _wallets: HashMap::new(),
        }
    }

    /// Create a new wallet
    pub async fn create_wallet(&self, password: &str, name: Option<&str>) -> SDKResult<WalletInfo> {
        // Generate new keypair
        let mut rng = rand::rngs::OsRng;
        let mut secret_bytes = [0u8; 32];
        rng.fill_bytes(&mut secret_bytes);
        let signing_key = SigningKey::from_bytes(&secret_bytes);
        let verifying_key = signing_key.verifying_key();

        // Generate wallet address
        let address = self.generate_address(&verifying_key);

        // Encrypt private key
        let encrypted_private_key = self.encrypt_private_key(&signing_key.to_bytes(), password)?;

        // Create wallet data
        let wallet_data = WalletData {
            address: address.clone(),
            name: name.map(|s| s.to_string()),
            encrypted_private_key,
            public_key: verifying_key.to_bytes().to_vec(),
            created_at: chrono::Utc::now().timestamp(),
            balance: 0.0,
        };

        // Store wallet (in a real implementation, this would be persisted)
        // For now, we'll just return the wallet info
        let wallet_info = WalletInfo {
            address,
            name: name.map(|s| s.to_string()),
            balance: 0.0,
            created_at: wallet_data.created_at,
        };

        Ok(wallet_info)
    }

    /// Import wallet from private key
    pub async fn import_wallet(&self, private_key_hex: &str, password: &str, name: Option<&str>) -> SDKResult<WalletInfo> {
        // Decode private key
        let private_key_bytes = hex::decode(private_key_hex)
            .map_err(|_| SDKError::InvalidInput("Invalid private key format".to_string()))?;

        if private_key_bytes.len() != 32 {
            return Err(SDKError::InvalidInput("Invalid private key length".to_string()));
        }

        // Create signing key
        let signing_key = SigningKey::from_bytes(&private_key_bytes.try_into().unwrap());
        let verifying_key = signing_key.verifying_key();

        // Generate wallet address
        let address = self.generate_address(&verifying_key);

        // Encrypt private key
        let encrypted_private_key = self.encrypt_private_key(&signing_key.to_bytes(), password)?;

        // Create wallet data
        let wallet_data = WalletData {
            address: address.clone(),
            name: name.map(|s| s.to_string()),
            encrypted_private_key,
            public_key: verifying_key.to_bytes().to_vec(),
            created_at: chrono::Utc::now().timestamp(),
            balance: 0.0,
        };

        // Store wallet (in a real implementation, this would be persisted)
        let wallet_info = WalletInfo {
            address,
            name: name.map(|s| s.to_string()),
            balance: 0.0,
            created_at: wallet_data.created_at,
        };

        Ok(wallet_info)
    }

    /// Get wallet information
    pub async fn get_wallet(&self, address: &str) -> SDKResult<WalletInfo> {
        // In a real implementation, this would load from storage
        // For now, we'll return a mock wallet
        Ok(WalletInfo {
            address: address.to_string(),
            name: Some("Demo Wallet".to_string()),
            balance: 1000.0,
            created_at: chrono::Utc::now().timestamp(),
        })
    }

    /// List all wallets
    pub async fn list_wallets(&self) -> SDKResult<Vec<WalletInfo>> {
        // In a real implementation, this would load from storage
        // For now, we'll return mock wallets
        Ok(vec![
            WalletInfo {
                address: "alice".to_string(),
                name: Some("Alice's Wallet".to_string()),
                balance: 1000.0,
                created_at: chrono::Utc::now().timestamp(),
            },
            WalletInfo {
                address: "bob".to_string(),
                name: Some("Bob's Wallet".to_string()),
                balance: 500.0,
                created_at: chrono::Utc::now().timestamp(),
            },
        ])
    }

    /// Sign transaction data
    pub async fn sign_transaction(&self, _address: &str, _password: &str, transaction_data: &[u8]) -> SDKResult<Vec<u8>> {
        // In a real implementation, this would:
        // 1. Load the wallet
        // 2. Decrypt the private key
        // 3. Sign the transaction data
        // For now, we'll return a mock signature
        let mut signature = Vec::new();
        signature.extend_from_slice(transaction_data);
        signature.extend_from_slice(b"mock_signature");
        Ok(signature)
    }

    /// Export private key
    pub async fn export_private_key(&self, _address: &str, _password: &str) -> SDKResult<String> {
        // In a real implementation, this would:
        // 1. Load the wallet
        // 2. Verify the password
        // 3. Decrypt and return the private key
        // For now, we'll return a mock private key
        Ok("mock_private_key_hex".to_string())
    }

    /// Update wallet balance
    pub async fn update_balance(&self, _address: &str, _balance: f64) -> SDKResult<()> {
        // In a real implementation, this would update the stored wallet data
        // For now, we'll just return success
        Ok(())
    }

    /// Delete wallet
    pub async fn delete_wallet(&self, _address: &str, _password: &str) -> SDKResult<()> {
        // In a real implementation, this would:
        // 1. Verify the password
        // 2. Remove the wallet from storage
        // For now, we'll just return success
        Ok(())
    }

    /// Generate wallet address from public key
    fn generate_address(&self, public_key: &VerifyingKey) -> String {
        let public_key_bytes = public_key.to_bytes();
        let mut hasher = sha2::Sha256::new();
        hasher.update(public_key_bytes);
        let hash = hasher.finalize();
        
        // Take first 20 bytes for address
        let address_bytes = &hash[..20];
        hex::encode(address_bytes)
    }

    /// Encrypt private key with password
    fn encrypt_private_key(&self, private_key: &[u8], password: &str) -> SDKResult<Vec<u8>> {
        // In a real implementation, this would use proper encryption
        // For now, we'll just encode it with the password
        let mut encrypted = Vec::new();
        encrypted.extend_from_slice(password.as_bytes());
        encrypted.extend_from_slice(private_key);
        Ok(encrypted)
    }

    /// Decrypt private key with password
    #[allow(dead_code)]
    fn decrypt_private_key(&self, encrypted_data: &[u8], password: &str) -> SDKResult<Vec<u8>> {
        // In a real implementation, this would use proper decryption
        // For now, we'll just extract the private key
        if encrypted_data.len() < password.len() {
            return Err(SDKError::WalletError("Invalid encrypted data".to_string()));
        }

        let password_bytes = password.as_bytes();
        if &encrypted_data[..password_bytes.len()] != password_bytes {
            return Err(SDKError::AuthError("Invalid password".to_string()));
        }

        Ok(encrypted_data[password_bytes.len()..].to_vec())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_wallet_creation() {
        let config = crate::SDKConfig::default();
        let wallet_manager = WalletManager::new(config);
        
        let wallet = wallet_manager.create_wallet("test_password", Some("test_wallet")).await.unwrap();
        assert_eq!(wallet.name, Some("test_wallet".to_string()));
        assert!(!wallet.address.is_empty());
    }

    #[tokio::test]
    async fn test_wallet_import() {
        let config = crate::SDKConfig::default();
        let wallet_manager = WalletManager::new(config);
        
        // Generate a test keypair
        let mut secret_bytes = [0u8; 32];
        rand::rngs::OsRng.fill_bytes(&mut secret_bytes);
        let _signing_key = ed25519_dalek::SigningKey::from_bytes(&secret_bytes);
        let private_key_hex = hex::encode(secret_bytes);
        
        let wallet = wallet_manager.import_wallet(&private_key_hex, "test_password", Some("imported_wallet")).await.unwrap();
        assert_eq!(wallet.name, Some("imported_wallet".to_string()));
    }

    #[tokio::test]
    async fn test_wallet_signing() {
        let config = crate::SDKConfig::default();
        let wallet_manager = WalletManager::new(config);
        
        let wallet = wallet_manager.create_wallet("test_password", None).await.unwrap();
        let transaction_data = b"test transaction data";
        
        let signature = wallet_manager.sign_transaction(&wallet.address, "test_password", transaction_data).await.unwrap();
        assert!(!signature.is_empty());
    }
}
